src/hotspot/share/gc/shenandoah/shenandoahControlThread.cpp

```diff
@@ -254,16 +244,16 @@ void ShenandoahControlThread::run_service() {
       }
     }
 
-    // Blow all soft references on this cycle, if handling allocation failure,
-    // either implicit or explicit GC request, or we are requested to do so unconditionally.
-    if (generation == select_global_generation() && (alloc_failure_pending || implicit_gc_requested || explicit_gc_requested || ShenandoahAlwaysClearSoftRefs)) {
-      heap->soft_ref_policy()->set_should_clear_all_soft_refs(true);
-    }
-
     bool gc_requested = (gc_mode() != none);
-    assert (!gc_requested || cause != GCCause::_last_gc_cause, "GC cause should be set");
+    assert (!gc_requested || cause != GCCause::_no_gc, "GC cause should be set");
 
     if (gc_requested) {
+      // Blow away all soft references on this cycle, if handling allocation failure,
+      // either implicit or explicit GC request, or we are requested to do so unconditionally.
+      if (generation == select_global_generation() && (alloc_failure_pending || implicit_gc_requested || explicit_gc_requested || ShenandoahAlwaysClearSoftRefs)) {
```

Comment:
Not a change you made, but why do we clear all soft refs if it's an implicit gc? Is memory pressure supposed to be very high here?

Reply:
Thanks!